src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAOQuerySupport.scala

```diff
@@ -0,0 +1,124 @@
+    }
+    ESShould(clauses)
+  }
+
+  def createQuery(criteria: LibrarySearchParams): Seq[QueryMap] = {
+    (criteria.searchString, criteria.searchFields.size) match {
+      case (None | Some(""), 0) => Seq(new ESMatchAll)
+      case (None | Some(""), _) => createListOfMusts(criteria.searchFields)
+      case (Some(searchTerm: String), 0) => Seq(new ESMatch(searchTerm.toLowerCase))
+      case (Some(searchTerm: String), _) => createListOfMusts(criteria.searchFields) :+ new ESMatch(searchTerm.toLowerCase)
+    }
+  }
+
+  def addAggregationsToQuery(searchReq: SearchRequestBuilder, aggFields: Seq[String], maxAggs: Option[Int]): SearchRequestBuilder = {
+    aggFields foreach { field: String =>
+      val terms = AggregationBuilders.terms(field)
+      if (maxAggs.isDefined) {
+        terms.size(maxAggs.get)
+      }
+      searchReq.addAggregation(terms.field(field + ".raw"))
+    }
+    searchReq
+  }
+
+  def createESSearchRequest(client: TransportClient, indexname: String, qmseq: Seq[QueryMap], from: Int, size: Int): SearchRequestBuilder = {
+    client.prepareSearch(indexname)
+      .setQuery(ESConstantScore(ESFilter(ESBool(ESMust(qmseq)))).toJson.compactPrint)
+      .setFrom(from)
+      .setSize(size)
+  }
+
+  def buildMainQuery(client: TransportClient, indexname: String, baseQuery: Seq[QueryMap], criteria: LibrarySearchParams): SearchRequestBuilder = {
+    val mainQuery = createESSearchRequest(client, indexname, baseQuery, criteria.from, criteria.size)
+    if (criteria.fieldAggregations.nonEmpty && criteria.fieldAggregations.size != criteria.searchFields.size) {
+      // if we are not collecting aggregation data (in the case of pagination)
+      // and if there are some aggregations that are not part of the search criteria
+      // then the aggregation data for those fields will be accurate from the main search
+      addAggregationsToQuery(mainQuery, criteria.fieldAggregations.diff(criteria.searchFields.keySet.toIndexedSeq), criteria.maxAggregations)
+    }
+    mainQuery
+  }
+
+
+  def buildAggregateQueries(client: TransportClient, indexname: String, baseQuery: Seq[QueryMap], criteria: LibrarySearchParams): Seq[SearchRequestBuilder] = {
+    // for aggregtions fields that are part of the current search criteria, we need to do a separate
+    // aggregate request *without* that term in the search criteria
+    (criteria.searchFields.keys map { field =>
+      val query = baseQuery.filterNot((m: QueryMap) => m.isInstanceOf[ESBool] && m.asInstanceOf[ESBool].bool.isInstanceOf[ESShould]
+        && m.asInstanceOf[ESBool].bool.asInstanceOf[ESShould].should.last.isInstanceOf[ESTerm]
+        && m.asInstanceOf[ESBool].bool.asInstanceOf[ESShould].should.last.asInstanceOf[ESTerm].term.last._1.equals(field))
+      // setting size to 0, we will ignore the actual search results
+      addAggregationsToQuery(createESSearchRequest(client, indexname, query, 0, 0), Seq(field), criteria.maxAggregations)
+    }).toSeq
+  }
```

Comment:
how about this?
```
  def buildAggregateQueries(client: TransportClient, indexname: String, baseQuery: Seq[QueryMap], criteria: LibrarySearchParams): Seq[SearchRequestBuilder] = {
    // for aggregations fields that are part of the current search criteria, we need to do a separate
    // aggregate request *without* that term in the search criteria
    (criteria.fieldAggregations intersect criteria.searchFields.keySet.toSeq) map { field =>
      val query = createQuery(criteria.copy(searchFields = criteria.searchFields - field))
      // setting size to 0, we will ignore the actual search results
      addAggregationsToQuery(createESSearchRequest(client, indexname, query, 0, 0), Seq(field), criteria.maxAggregations)
    }
  }
```

I've done two things:
1. with the initial `criteria.fieldAggregations intersect`, I'm checking to make sure we're requesting an aggregation for any given field
2. by reusing `createQuery`, simplify the body of the function quite a bit